tests/ssg_test_suite/rule.py

```diff
@@ -261,17 +247,74 @@ def _prepare_environment(self, scenarios_by_rule):
+
+        Returns:
+            Named tuple Rule having these fields:
+                directory -- absolute path to the rule "tests" subdirectory
+                            containing the test scenarios in Bash
+                id -- full rule id as it is present in datastream
+                short_id -- short rule ID, the same as basename of the directory
+                            containing the test scenarios in Bash
+                template -- name of the template the rule uses
+                local_env_yaml -- env_yaml specific to rule's own context
+                rule -- rule class, contains information parsed from rule.yml
+        """
+
+        # Here we need to perform some magic to handle parsing the rule (from a
+        # product perspective) and loading any templated tests. In particular,
+        # identifying which tests to potentially run involves invoking the
+        # templating engine.
+        #
+        # Begin by loading context about our execution environment, if any.
+        product_yaml = common.get_product_context(product)
+
+        for dirpath, dirnames, filenames in common.walk_through_benchmark_dirs(
+                product):
+            if not common.is_rule_dir(dirpath):
+                continue
+            short_rule_id = os.path.basename(dirpath)
+            full_rule_id = OSCAP_RULE + short_rule_id
+            if not self._rule_should_be_tested(short_rule_id, target):
                 continue
             if not xml_operations.find_rule_in_benchmark(
-                    self.datastream, self.benchmark_id, rule.id):
+                    self.datastream, self.benchmark_id, full_rule_id):
                 logging.error(
                     "Rule '{0}' isn't present in benchmark '{1}' in '{2}'"
-                    .format(rule.id, self.benchmark_id, self.datastream))
+                    .format(full_rule_id, self.benchmark_id, self.datastream))
+                continue
+
+            # Load the rule itself to check for a template.
+            rule, local_env_yaml = common.load_rule_and_env(
+                dirpath, product_yaml, product)
+
+            # Before we get too far, we wish to search the rule YAML to see if
+            # it is applicable to the current product. If we have a product
+            # and the rule isn't applicable for the product, there's no point
+            # in continuing with the rest of the loading. This should speed up
+            # the loading of the templated tests. Note that we've already
+            # parsed the prodtype into local_env_yaml
+            if product and local_env_yaml['products']:
+                prodtypes = local_env_yaml['products']
+                if "all" not in prodtypes and product not in prodtypes:
+                    continue
+
+            tests_dir = os.path.join(dirpath, "tests")
+            template_name = None
+            if rule.template and rule.template['vars']:
+                template_name = rule.template['name']
```

Comment:
I noticed that there is a very similar code in the function `template_rule_tests` and there it also loads rules from rule.yml. But the `template_rule_tests`  is called during tarball creation.

I'm not mentioning the code duplication because what is scary is that `template_rule_tests` is called on all rule directories in linux_os tree therefore all rule.yamls (thousands) are parsed and loaded for the second time (!), even if you test a single rule or if you test a rule that doesn't exists. It's done at the moment when we already know which rules and scenarios will be executed.

FTR the call trace just to see where  `template_rule_tests` is called, notice that `self._prepare_environment(scenarios_by_rule_id)` is there which means it's called after we already loaded the rules and scenarios we need.
```
  File "/home/jcerny/work/git/scap-security-guide/tests/test_suite.py", line 462, in <module>
    main()
  File "/home/jcerny/work/git/scap-security-guide/tests/test_suite.py", line 458, in main
    options.func(options)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/rule.py", line 620, in perform_rule_check
    checker.test_target(options.target)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/oscap.py", line 699, in test_target
    self._test_target(target)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/rule.py", line 423, in _test_target
    self._prepare_environment(scenarios_by_rule_id)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/rule.py", line 248, in _prepare_environment
    self.remote_dir = common.send_scripts(self.test_env)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/common.py", line 495, in send_scripts
    archive_file = create_tarball(test_env.product)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/common.py", line 463, in create_tarball
    templated_tests = template_tests(product=product)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/common.py", line 453, in template_tests
    raise exp
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/common.py", line 450, in template_tests
    template_rule_tests(product, product_yaml, template_builder, tmpdir, dirpath)
  File "/home/jcerny/work/git/scap-security-guide/tests/ssg_test_suite/common.py", line 386, in template_rule_tests
```

Reply:
@matejak  I have transformed this problem into an issue https://github.com/ComplianceAsCode/content/issues/8765 because solving this problem here would make this PR even larger.

Reply:
Yes, I completely agree with the description of the issue.